docs/design/arbiter.md

```diff
@@ -0,0 +1,191 @@
+// get an initial list of devices where bricks can be placed
+// as well as converting device IDs to device entry objects.
+// The idea is to keep db/connection/caching logic away from
+// the placement algorithms in the placer interface.
+type DeviceSource interface {
+	// Devices returns a list of all the Device entries that can
+	// be considered for the upcoming brick placement.
+	Devices() ([]*DeviceEntry, error)
+	// Device looks up a device id and resolves it to a Device
+	// entry object.
+	Device(id string) (*DeviceEntry, error)
+}
+
+// PlacementOpts is interface that is meant for passing the
+// somewhat complex set of options needed by the placer code
+// and hiding the sources of the these values away from the
+// placer implementations.
+type PlacementOpts interface {
+	// BrickSizes returns values needed to calculate the
+	// size of the brick on disk.
+	BrickSizes() (uint64, float64)
+	// BrickOwner returns the ID of object that will "own" the brick
+	BrickOwner() string
+	// BrickGid return the ID of the GID the brick will use
+	BrickGid() int
+	// SetSize returns the size of the Brick Sets that will be
+	// allocated.
+	SetSize() uint
+	// SetCount returns the total number of Brick Sets that
+	// will be produced.
+	SetCount() uint
+	// ValidDevice is a predicate function that the higher level
+	// can define in order to cause the placement algorithm to
+	// exlucde certain devices.
+	ValidDevice(*BrickSet, *DeviceEntry) bool
+}
+
+// BrickPlacer implementations take their source devices and
+// options and place new bricks on devices (if possible).
+// The exact placement depends on the implementation and
+// the input options.
+type BrickPlacer interface {
+	// PlaceAll constructs a full sequence of brick sets and
+	// corresponding device sets for those bricks.
+	PlaceAll(DeviceSource, PlacementOpts) (
+		BrickAllocation, error)
+
+	// Replace constructs a brick allocation constrained to
+	// a single brick set where the brick set is already populated
+	// but a brick with the given index into the set needs to
+	// be replaced.
+	Replace(DeviceSource, PlacementOpts, BrickSet, int) (
+		BrickAllocation, error)
+}
+```
+
+The final form of these interfaces is still subject to change but
+this is a more-than-rough sketch of interfaces that should work
+for both arbiter and non-arbiter volumes for the volume
+create, volume expand, and brick replace use cases.
+
+
+## Need to properly calculate arbiter brick sizes
+
+Arbiter bricks are expected to be smaller than normal bricks as they only store
+volume metadata. Because of this and that one of the selling points of arbiter
+volumes is is space saving (and thus cost saving) feature we need to size the
+arbiter bricks smaller than the data bricks but large enough to hold all the 
+metadata needed.
+
+
+## Touchpoints
+
+This is a summary of the areas of the code that will need to be modified
+to support arbiter volumes:
+
+* Volume Create
```

Comment:
may be we can mention the key `arbiter` on volume options in heketi cli or api client can trigger these kind of volumes.